storage/repository/repository.go

```diff
@@ -0,0 +1,198 @@
+}
+
+// WriteableRepository is an extension of the Repository interface which also
+// supports obtaining a writer for URIs of the scheme it is registered to.
+//
+// Since 2.0.0
+type WriteableRepository interface {
+	Repository
+
+	// Writer will be used to implement calls to storage.WriterTo() for
+	// the registered scheme of this repository.
+	//
+	// Since 2.0.0
+	Writer(u fyne.URI) (fyne.URIWriteCloser, error)
+
+	// CanWrite will be used to implement calls to storage.CanWrite() for the
+	// registered scheme of this repository.
+	//
+	// Since 2.0.0
+	CanWrite(u fyne.URI) (bool, error)
+
+	// Delete will be used to implement calls to storage.Delete() for the
+	// registered scheme of this repository.
+	//
+	// Since 2.0.0
+	Delete(u fyne.URI) error
+}
+
+// ListableRepository is an extension of the Repository interface which also
+// supports obtaining directory listings (generally analogous to a directory
+// listing) for URIs of the scheme it is registered to.
+//
+// Since 2.0.0
+type ListableRepository interface {
+	Repository
+
+	// CanList will be used to implement calls to storage.Listable() for
+	// the registered scheme of this repository.
+	//
+	// Since 2.0.0
+	CanList(u fyne.URI) (bool, error)
+
+	// List will be used to implement calls to storage.List() for the
+	// registered scheme of this repository.
+	//
+	// Since 2.0.0
+	List(u fyne.URI) ([]fyne.URI, error)
+}
+
+// HierarchicalRepository is an extension of the Repository interface which
+// also supports determining the parent and child items of a URI.
+//
+// Since 2.0.0
+type HierarchicalRepository interface {
+	Repository
+
+	// Parent will be used to implement calls to storage.Parent() for the
+	// registered scheme of this repository.
+	//
+	// A generic implementation is provided in GenericParent(), which
+	// is based on the RFC3986 definition of a URI parent.
+	//
+	// Since 2.0.0
+	Parent(fyne.URI) (fyne.URI, error)
+
+	// Child will be used to implement calls to storage.Child() for
+	// the registered scheme of this repository.
+	//
+	// A generic implementation is provided in GenericParent(), which
+	// is based on RFC3986.
+	//
+	// Since 2.0.0
+	Child(fyne.URI) (fyne.URI, error)
+}
+
+// CopyableRepository is an extension of the Repository interface which also
+// supports copying referenced resources from one URI to another.
+type CopyableRepository interface {
+	Repository
+
+	// Copy will be used to implement calls to storage.Copy() for the
+	// registered scheme of this repository.
+	//
+	// A generic implementation is provided by GenericCopy().
+	//
+	// NOTE: the first parameter is the source, the second is the
+	// destination.
+	//
+	// NOTE: if storage.Copy() is given two URIs of different schemes, it
+	// is possible that only the source URI will be of the type this
+	// repository is registered to handle. In such cases, implementations
+	// are suggested to fail-over to GenericCopy().
+	//
+	// Since 2.0.0
+	Copy(fyne.URI, fyne.URI) error
```

Comment:
Should this return the number of bytes copied? Ditto for Move below, the package-level functions, and their Generic counterparts.

Reply:
I don't think this makese sense. Move and Copy are file operations. it either succeeds (file copied) or it errors (file not copied).

Reply:
I think that's a good idea, people can always discard it if they want to.

Reply:
If you were writing a stream handling API or some other thing where you were dealing with the content maybe. The only API we have that touches the content of a file is Read/Write. That is a pretty clear separation.

Reply:
If you go down this strange path then Delete() should return the bytes freed as well.

Reply:
Trying to find examples but Java does not have file copying. Apple's SDK is a high level file copy, no mention of bytes. .Net similarly does not provide raw data information in these high level APIs.

Reply:
I'm not convinced. Repositories provide io.Readers/io.Writers for a URI, Copy is an operation from Read (which returns count) to Write (which also returns count). Even https://golang.org/pkg/io/#Copy returns count.

If an implementer has the data, and in most cases they will, they should be able to return it as callers may want it, and in the rare case the implementer doesn't or can't know, they can just return 0.

I don't think Delete needs to return the bytes freed, but I wouldn't be opposed.

Reply:
> Even https://golang.org/pkg/io/#Copy returns count.

`io.Copy`, as per it's location, is an IO operation not a file abstraction.
A copy  *may* read a reader and write to a writer, but we hope not on optimal Repos, this it is an implementation detail that should not leak out of an abstraction.

Reply:
Actually, I think I am now opposed for the reason @andydotxyz said. In an optimized implementation, the back-end may not know the number of bytes copied, and determining this may increase complexity / reduce performance.

Reply:
Can you give an example of where that may be the case?

I think it is incorrect to think of this a file abstraction as it was a conscious effort to use URIs instead of files.

Reply:
Any copy that issues the "copy x location to y location" will not know how many bytes. The unoptimised "copy x's bytes from repo1 into y on repo2" will know - but what matters is whether it succeeds or not.
When Read returns the count we have access to the bytes so it is meaningful metadata. When copying we have no access to the bytes at this API level so even if it was calculable it would relate to any handle we have.

Reply:
For example, in some hypothetical COW-style system, a copy operation may simple involve creating additional references to specific filesystem blocks without actually duplicating the data on them. One could imagine something similar for references into some kind of database or file archive.

Reply:
To flip the discussion around, what is the use case of supporting this? I struggle to think of a context where this would be actionable information.

Reply:
I do not agree with adding the bytes to either `Read`, `Write`, `Copy`, or `Move`.

The only reason I would be for this, is if our `Read` and `Write` matched `io.Reader` and `io.Writer`.

However, since we are dealing in URIs in various formats, this is not feasible.

Reply:
the Read() returns an extension of `io.ReadCloser`, it is not actually doing any reading - thanks for prodding me on this @albinoGeek, I had forgotten that.

Reply:
ack